core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -837,17 +847,31 @@ private[spark] class TaskSchedulerImpl(
   }
 
   /**
-   * Marks the task has completed in all TaskSetManagers for the given stage.
+   * Marks the task has completed in all TaskSetManagers(active / zombie) for the given stage.
    *
    * After stage failure and retry, there may be multiple TaskSetManagers for the stage.
    * If an earlier attempt of a stage completes a task, we should ensure that the later attempts
    * do not also submit those same tasks.  That also means that a task completion from an earlier
    * attempt can lead to the entire stage getting marked as successful.
+   * And there's a situation that the active TaskSetManager corresponding to the stage may
+   * haven't been created at the time we call this method. And it is possible since the behaviour
+   * of calling on this method and creating active TaskSetManager is from two different threads,
+   * which are "task-result-getter" and "dag-scheduler-event-loop" separately. Consequently, under
+   * this situation, the active TaskSetManager which is created later could not learn about the
+   * finished partitions and keep on launching duplicate tasks, which may lead to job fail for some
+   * severe cases, see SPARK-25250 for details. So, to avoid the problem, we record the finished
+   * partitions for that stage here and exclude the already finished tasks when we creating active
+   * TaskSetManagers later by looking into stageIdToFinishedPartitions. Thus, active TaskSetManager
+   * could be always notified about the finished partitions whether it has been created or not at
+   * the time we call this method.
```

Comment:
This comment is good, but a little too wordy.  I'd cut it down to something like:

There is also the possibility that the DAGScheduler submits another taskset at the same time as we're marking a task completed here -- that taskset would have a task for a partition that was already completed.  We maintain the set of finished partitions in stageIdToFinishedPartitions, protected by this, so we can filter those tasks when the taskset is submitted.  See SPARK-25250 for more details.

Reply:
And can you also add a comment that this should only be called with a lock on `this`?

Reply:
updated.